A31-xds-routeaction-support.md

```diff
@@ -0,0 +1,214 @@
+Config](https://github.com/grpc/grpc/blob/master/doc/service_config.md),
+provided by the Name Resolver, and currently has limited matching semantics
+(per-service or per-method with exact name matching only).
+
+In xDS, RDS
+[provides](https://www.envoyproxy.io/docs/envoy/latest/api-v2/api/v2/route/route_components.proto#route-routeaction)
+the ability to configure timeouts and several other features using its matchers,
+which support a more rich set of matching semantics than gRPC, including
+examining the outgoing HTTP headers and using regexes.
+
+Further complicating matters, gRPC needs to be able to determine timeout
+settings before invoking the Load Balancing Policy, which is the point at which
+the bulk of the current xDS support is integrated with gRPC.
+
+Extending gRPC to support these matchers directly is not a preferred solution to
+the problem, as it could impact performance for all gRPC users, increase our API
+surface, and grow our code base for features we otherwise do not intend to
+implement.
+
+
+## Proposal
+
+This proposal is broken up into two sections: the addition of new functionality
+to gRPC itself, and then a description of using that functionality to implement
+support for xDS RouteActions.
+
+### New Functionality in gRPC
+
+In order to allow the xDS matchers to configure gRPC behavior, but to avoid
+implementing support for them directly in gRPC, a new "per-RPC Config Selector"
+will be added for internal-only use.  This will be provided by the Name Resolver
+component when calling `Update` on the channel in a way that does not extend our
+publicly-visible API, similarly to how the xDS Client is provided to the Load
+Balancing Policy.
+
+If present, this Config Selector will be invoked once at the beginning of every
+RPC (after the name resolver has produced a result), and it will produce four
+things:
+
+1.  The same per-RPC configuration which is currently determined by the Service
+    Config's MethodConfig.  A default implementation will return the Service
+    Config's selected values.
+1.  An optional interceptor or callback.  The interceptor would be invoked by
+    the channel immediately for the RPC, or the callback would be invoked at
+    some point after the channel knows the RPC will no longer call the Load
+    Balancing Policy's Pick method and before the end of the RPC.  This is
+    necessary for cluster lifecycle management, which will be explained in
+    detail later.
+1.  Information which will be passed to the Load Balancing Policy's Pick method
+    by the channel.  E.g. in Go this could be done via the Context, and Java may
+    use CallOptions.  If an interceptor is used for (2) it is possible that the
+    interceptor may be capable of satisfying this requirement directly.
+1.  An RPC status; if not OK, the RPC is terminated with this status.  This is
+    needed in case the RPC cannot be routed.  Like (3), this could be satisfied
+    via an interceptor.
```

Comment:
Can delay injection be supported by Config Selector for supporting fault injection? I might be wrong here. Delay injection is tightly coupled with threading model. For Golang, we could possibly block the `selectConfig` method. But for Java and Core, it seems the `getCallOptions` or `GetCallConfig` won't allow blocking or be scheduled asynchronously.

Is it possible to consider an alternative design that uses callback instead of directly return?

Or this functionality must be implemented as an interceptor? In that case, what's our plan to support interceptors in Config Selector across languages?

Reply:
I think we should discuss fault injection separately -- I don't think that's something that should be bundled with the ConfigSelector functionality.

In Java and Go, it wouldn't be that hard to have the ConfigSelector return an interceptor -- in fact, that may be how we wind up implementing it in those languages.  But there is no reasonable way to do anything like that in C-core today.

This is a problem that we are going to have to solve eventually, because there will be use cases where people will want to configure third-party interceptors via xDS.  However, we don't currently have a good solution for that problem today, and figuring out how to do it in C-core is going to be a lot of work.  I don't think we want the fault injection design to depend on that.

Let's chat more about this offline.  I suspect that in our current architecture, we may need to build support for this directly into core, just like we've (partially) done for retries.

Reply:
Thanks for the helpful info. I will read about retry mechanism implementation next before chat with you.